node/modules/storageminer.go

```diff
@@ -351,6 +353,48 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
 	return sst, nil
 }
 
+func F3Participation(mctx helpers.MetricsCtx, lc fx.Lifecycle, api v1api.FullNode, minerAddress dtypes.MinerAddress) error {
+	ctx := helpers.LifecycleCtx(mctx, lc)
+	b := &backoff.Backoff{
+		Min:    1 * time.Second,
+		Max:    1 * time.Minute,
+		Factor: 1.5,
+		Jitter: false,
+	}
+	go func() {
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+			ch, err := api.F3Participate(ctx, address.Address(minerAddress))
+
+			if errors.Is(err, context.Canceled) {
+				log.Errorf("Context cancelled while attampting F3 participation: %+v", err)
+				return
+			} else if errors.Is(err, full.ErrF3Disabled) {
+				log.Errorf("Cannot participate in F3 as it is disabled: %+v", err)
+				return
+			} else if err != nil {
+				log.Errorf("while starting to participate in F3: %+v", err)
+				// use exponential backoff to avoid hotloop
+				time.Sleep(b.Duration())
+				continue
+			}
+
+			for err := range ch {
```

Comment:
Put the for loop in a final else clause and combine the sleep statements?

Reply:
I have reworked the logic here to what i believe is the intent. I'd be grateful if you give it another look please.